### DISCUSSION TITLE
Map v/s Scan based Loop Closures Comparison

### DISCUSSION THREAD

**Discussion by LL884800** [PRBonn/MapClosures#37] [General]
Hello, I want to know how I can be sure that the loop diagram is obtained by using this algorithm. Then how can I be sure that the robot has returned to its original position? If I only get the loop diagram, how can I compare it with the scan context? After all, scan context It is possible to obtain the specific pose of the robot when it returns to its original position.

**Comment by saurabh1002** [comment]
Hello @LL884800, thanks for showing interest in our work.
To begin with, I am not exactly sure what you mean by a "loop diagram". It would be great if you can elaborate a bit further!

Secondly, a one-to-one comparison with ScanContext is not possible since the two approaches are different in the way they process data. ScanContext works on individual scans, whereas we work on local maps generated from registered scans within a certain distance. So each local map contains multiple individual scans.

As a result the start position of the robot and certain number of subsequent scans are contained within the first local map. To figure out if you are revisiting the start location, you can check for the local map index of the loop closures obtained, whether the reference index is `0` or not.

The following image shows how we save the closure information in out `pipeline.py`:
![image](https://github.com/user-attachments/assets/f1b1769b-78c3-4d6f-9d64-d9ed76f4468c)

The relative pose we provide is between the first scans in the source and the target local maps respectively. This is because each local map is in the local reference frame of the first scan it contains.

**Comment by saurabh1002** [comment]
You can check out #24 to see how we integrate the detected closures between local maps to perform a pose-graph optimization. This PR will be soon merged with the main branch.

**Comment by saurabh1002** [comment]
Hey, it would be great if you can add comments to this issue rather than editing the title of the issue in reply.

**Comment by saurabh1002** [comment]
You can find the way we compute scan-level closure indices from the map-level closures in `tools/evaluation.py` and is also mentioned in the paper. We use the provided script to compute scan-level closures to compare with scan context, through the Precion Recall and F1 score.

**Comment by LL884800** [comment]
Hello, my understanding is that the final result of the map closure algorithm is a possible loop graph index, and then an optimal loop index can be obtained based on the number of interior points. I can think that these two local maps contain the posture of the robot looping at the same position. points, but how can I determine which point is the same location? What I want to know here is how I determine which point is considered to have returned to the same position. The map closure algorithm is the loop closure between local maps, while the scanning context is the loop closure detection between single frames. I need to compare these two algorithms later. Compare how mapclousre algorithm loopback is performed? One is a partial image and the other is a single frame. How do they compare?

**Comment by tizianoGuadagnino** [comment]
@LL884800 I honestly do not understand what you mean. Can you give us a little bit more detail in what you want/what you want to do?

**Comment by LL884800** [comment]
In the map closure algorithm, how to accurately find the point where the robot returns to the same position from the matching of the local map, and effectively compare it with the loopback point detected through a single frame image in the scanning context algorithm

**Comment by saurabh1002** [comment]
I try to explain the same in my previous reply to the issue you opened [here](https://github.com/PRBonn/MapClosures/discussions/37#discussioncomment-10530859)
Since we provide a more general loop closure algorithm and not just a special case of loop closing at the same point where you start, using local maps, there is no explicit way to do what you want. However, it is easy to extract the needed information from our detected closure.

The necessary information is:
1. Each local map `Li` contains `Ni` consecutive scans depending on the travel distance obtained from LiDAR odometry, and is in the reference frame at the first scan in the local map.
2. When our algorithm says that it found a closure between local maps with map indices `ref` and `query`, the 2D pose estimate thereby provided is the relative transform between the first scans in the respective localmaps `L_ref` and `L_query`.
3. Once you apply this transform to the set of odometry poses comprising the reference local map `L_ref`, the subset of trajectory within the two local maps should be aligned already.
4. After step 3, you can simply use a distance threshold to check spatial closeness of any two scans from the loop closed local maps. Also, the relative transform between any two such scans, after alignment in 2D, should be corresponding to the loop closure estimate between those individual scans.
5. All of this can be found implemented [here](https://github.com/PRBonn/MapClosures/blob/799d141662f4f30e038bcc6ce6f52a4aab54ea5f/python/map_closures/tools/evaluation.py#L43), which we use to compare our approach with other single-scan approaches like the ScanContext.
6. This is also mentioned in the corresponding research paper [here](https://www.ipb.uni-bonn.de/pdfs/gupta2024icra.pdf)

Coming back to the very specific case you are interested in, which is detecting if the robot revisits the very specific start location again, you need the following:
1. First, our algorithm needs to give you a loop closure with reference local map id = 0, because the first local map also contains the first scan of the sequence. If this is not the case, then either there is no closure at that location, or our pipeline is not able to detect it. In the latter case, I would love to take a look at your data and figure out why it does not work, and try to fix our shortcomings.
2. If you are able to detect the case mentioned above, then you can just align the chunk of trajectories between the local maps using step 2-3 above.
3. After that you can just check for the closest scan in the query local map to the first scan in the reference map (which should be the first scan of the sequence).

I hope this was helpful. If not, I am happy to help you with what you want if there is no issue from your side to share your data.

**Reply by parasrdr** [reply]
Hi,

Thanks for the amazing work. As you explained in point 1, a local map Li accumulates consecutive scans depending on the travel distance. After reading paper and having looked at your pipeline code, I noted its the displacement of last scan to first scan (final to initial pose)  which is configured as (local map factor * max sensor range). In case the platform stays at same place or moves  in circles inside this distance, the  same local map generation will continue & it won't  move to next local map id.  have I understood correctly ?

**Reply by saurabh1002** [reply]
Hi,
Yes, you understood it right. This is because the odometry pipeline we use i.e. KISS-ICP, internally uses a local map-to-frame alignment, and its pose estimate will not drift significantly as long as you keep looping within a small area contained in your local map. KISS-ICP updates its local map in a moving window style based on the displacement as well.

**Reply by saurabh1002** [reply]
However, with the `local_map_factor` parameter we provide a way to still create local maps and detect closures with a small map as well, which might be necessary in smaller indoor environments.

**Reply by parasrdr** [reply]
Thanks for the reply.
